### ISSUE TITLE
parameter to control count of e.g. digits required

### ISSUE THREAD

**Issue by faxemaxee** [tarunbatra/password-validator#39]
Hey,

thanks for the great package, I was wondering what your thoughts are when it comes to enable something like this:

`schema.has().digits(3)` – which would check if there are at least 3 digits

I think we could achieve this using `.has(regex)` but this might be a nice addition.

What do you think?

Max

**Comment by tarunbatra** [comment]
Hi @faxemaxee, thanks for the suggestion! The idea has merit and is very doable.

Will explore this and update here. Needless to say PRs are also welcome.

**Comment by faxemaxee** [comment]
Nice, I'd be happy to contribute. My first thought was moving the regex in constants from "real" regex to string like this:

```
regex: {
    digits: '\d.*',
    letters: '[a-zA-Z].*',
    symbols: '[`~\!@#\$%\^\&\*\(\)\-_\=\+\[\{\}\]\\\|;:'",<.>\/\?€£¥₹].*',
    spaces: '[\s].*'
  }
```

and then build a proper regex inside the lib functions utilizing `string.repeat(count)` with min = 1.
AFAIK this would allow checking for multiple occurrences while allow those anywhere in the string.

We could/should also check for upper/lowercase using regex and the same logic.

Any thought/missed pitfalls regarding this brief description?

**Comment by tarunbatra** [comment]
Thanks for helping with this. IMO, augmenting base regex strings based on the `count` provided (if any), looks like a clean solution to me. Few points to consider:
1. Do we want to generate final regex of the type `(\d.*)(\d.*)(\d.*)` or `(\d.*){3,}`. Any performance trade-offs here should be considered.
2. `uppercase()` and `lowercase()` was moved away from regex implementations due to #23. Don't think `minCount` functionality can be supported in the current approach. Another trade-off here.

**Comment by faxemaxee** [comment]
Thanks for your reply, lets discuss this further.

1. I can not provide any insight on what options has better performance, any thoughts from your side?
2. I get the point regarding upper- and lowercase, I tried to think about a solution using the current approach, the only viable option I can think of would be looping through the letters one by one and doing the exact check. Something like:
```
while(thereIsANextLetter && (upperCaseLetters < requiredCountOfUppercaseLetters)) {
    if (currentLetter === currentLetter.toUpperCase()) {
     upperCaseLetters++
   }
}

this would, of course, lower performance. 

Any thoughts?

**Comment by tarunbatra** [comment]
Sorry for ghosting!

> looping through the letters one by one and doing the exact check

I think the approach proposed by you will serve well since it doesn't degrade the performance of existing features until something better can be thought of.

> it seems like the (\d.*){3,} approach uses less steps

This was my expectation too. Should go with this. Curious to know how did you test though?

**Comment by faxemaxee** [comment]
No worries!

Glad you're fine with the approach.

I used https://regex101.com/ which provides me with the amount of steps and ms needed. I then tested 20 random strings with both regex approaches, different lengths and different `requiredLinks` -> turned out that `(\d.*){3,}` was just faster with fewer steps.

**Comment by tarunbatra** [comment]
Yes, it seems to be a good indicator of performance. Thanks! Feel free to ask more questions, if needed at any point.

**Comment by faxemaxee** [comment]
@tarunbatra just pushed the PR, feedback welcome. Wasn't too sure about code style so just tell me if you want me to adjust anything!

**Comment by tarunbatra** [comment]
Thank you for the PR, @faxemaxee. I took a glimpse and it looked good. I will check it soon and get back to you.

**Comment by tarunbatra** [comment]
Resolved in [`v5.1.0`](https://tarunbatra.com/password-validator/5.1.0/). 👏
